src/sensor-fusion/fusion-pose-sensor.js

```diff
@@ -25,211 +30,280 @@ import * as Util from '../util.js';
+export default class FusionPoseSensor {
+  constructor(kFilter, predictionTime, yawOnly, isDebug) {
+    this.yawOnly = yawOnly;
+
+    this.accelerometer = new MathUtil.Vector3();
+    this.gyroscope = new MathUtil.Vector3();
+
+    this.filter = new ComplementaryFilter(kFilter, isDebug);
+    this.posePredictor = new PosePredictor(predictionTime, isDebug);
+
+    this.filterToWorldQ = new MathUtil.Quaternion();
+
+    // Set the filter to world transform, depending on OS.
+    if (Util.isIOS()) {
+      this.filterToWorldQ.setFromAxisAngle(new MathUtil.Vector3(1, 0, 0), Math.PI / 2);
+    } else {
+      this.filterToWorldQ.setFromAxisAngle(new MathUtil.Vector3(1, 0, 0), -Math.PI / 2);
+    }
+
+    this.inverseWorldToScreenQ = new MathUtil.Quaternion();
+    this.worldToScreenQ = new MathUtil.Quaternion();
+    this.originalPoseAdjustQ = new MathUtil.Quaternion();
+    this.originalPoseAdjustQ.setFromAxisAngle(new MathUtil.Vector3(0, 0, 1),
+                                             -window.orientation * Math.PI / 180);
+
+    this.setScreenTransform_();
+    // Adjust this filter for being in landscape mode.
+    if (Util.isLandscapeMode()) {
+      this.filterToWorldQ.multiply(this.inverseWorldToScreenQ);
+    }
+
+    // Keep track of a reset transform for resetSensor.
+    this.resetQ = new MathUtil.Quaternion();
+
+    this.isFirefoxAndroid = Util.isFirefoxAndroid();
+    this.isIOS = Util.isIOS();
+
+    this.orientationOut_ = new Float32Array(4);
+
+    // Store information on API in use (either 'devicemotion' or 'sensor'),
+    // as well as any errors observed from using sensors for debugging.
+    this.api = null;
+    this.error = null;
+
+    this.onDeviceMotion_ = this.onDeviceMotion_.bind(this);
+    this.onOrientationChange_ = this.onOrientationChange_.bind(this);
+    this.onMessage_ = this.onMessage_.bind(this);
+    this.onSensorError_ = this.onSensorError_.bind(this);
+    this.onGyroscopeRead_ = this.onGyroscopeRead_.bind(this);
+    this.onAccelerometerRead_ = this.onAccelerometerRead_.bind(this);
+
+    this.start();
   }
 
-  this.inverseWorldToScreenQ = new MathUtil.Quaternion();
-  this.worldToScreenQ = new MathUtil.Quaternion();
-  this.originalPoseAdjustQ = new MathUtil.Quaternion();
-  this.originalPoseAdjustQ.setFromAxisAngle(new MathUtil.Vector3(0, 0, 1),
-                                           -window.orientation * Math.PI / 180);
-
-  this.setScreenTransform_();
-  // Adjust this filter for being in landscape mode.
-  if (Util.isLandscapeMode()) {
-    this.filterToWorldQ.multiply(this.inverseWorldToScreenQ);
+  start() {
+    // Only listen for postMessages if we're in an iOS and embedded inside a cross
+    // origin IFrame. In this case, the polyfill can still work if the containing
+    // page sends synthetic devicemotion events. For an example of this, see
+    // the iframe example in the repo at `examples/iframe.html`
+    if (Util.isIOS() && Util.isInsideCrossOriginIFrame()) {
```

Comment:
Yeah the iframe usage concerns me the most in these changes. But in the future, we can mostly offload the permissions onto developers using iframes since they'll be required to set feature policies and the APIs will either "just work" (or "not work"), correct?